### PR TITLE
Fix script loading on Godot 4.3

### LINKED ISSUE #187: Pandora fails to assign script with compiled script files
Godot version: 4.3 beta2

**Describe the bug**
When exporting a game, [Godot 4.3 allows to choose in which mode to export the script files](https://godotengine.org/article/dev-snapshot-godot-4-3-beta-1/#gdscript).
When this options is set to compress, pandora breaks, ~as it tries to load a `.gd` file, that instead got converted in `.gdc`~.
After further investigation, it appears that the issue comes from the [`get_entity_class`](https://github.com/bitbrain/pandora/blob/godot-4.x/addons/pandora/util/script_util.gd#L40-L45) implementation, in particular `EntityClass.has_source_code()` returns false with compiled scripts.

**To Reproduce**
1. Try to call a PandoraEntity inherited script's method
2. Export a project with pandora
3. Lookup at the console file.
4. On builds with binary gdscript files (exported with GDScript mode set to Compressed binary token/Binary tokens), an error will be throwed.

**Expected behavior**
Pandora assigns the correct scripts and the method call executes as it should.

**Screenshots**

Error
Editor Screenshot
Game Console Screenshot

**Desktop (please complete the following information):**
 - OS: Windows 11

**Additional context**
[This ~might be~ is the affected file](https://github.com/bitbrain/pandora/blob/godot-4.x/addons/pandora/util/script_util.gd)

### LINKED ISSUE #196: Issues with Godot 4.3
Godot version: 4.3-beta3

**Describe the bug**
Refer to godotengine/godot#94150 for further details

TL;DR:
The type handler code uses `Script.has_source_code` to test if a script is valid.
This is `false` for scripts with binary tokenization. `Script.can_instantiate` should be used.

The MRP from the orginal issue does work with this change, but fails when copying the newest version of pandora in there and making the change.

I found that when running the game, data.pandora gets saved in a corrupted way:
{"type": "undefined", "value": "<CompressedTexture#84384>"} where it should be {"type": "resource", "value": "res://godot.png"}. As stated this does not happen with the pandora version from the MRP, I don't know which version that is though.

**To Reproduce**
- Use the MRP from godotengine/godot#94150
- Follow the steps there to observe the issue with binary tokenization
- exchange `has_source_code` with `can_instantiate` to see that it fixes the issue
- copy the newest version of pandora to the addons folder
- exchange `has_source_code` with `can_instantiate` to see that it does not work anymore
- observe `data.pandora` to see that it get's corrupted when running the game

**Expected behavior**
- the texture should load when run through the editor and when exported with binary tokenization

**Screenshots**
- see godotengine/godot#94150

**Desktop (please complete the following information):**
 - OS: Fedora
 - Version 40

### PR DESCRIPTION
**Pull request by shomykohai**
## Description

Fixes various issue that came from Godot 4.3 changes in GDScript and loading.
The key point to solve this issue has been found thanks to the discussion on godotengine/godot#94150 and with the help of @HolonProduction.

Supersedes #189.

## Addressed issues

closes bitbrain/pandora#187
closes bitbrain/pandora#196

### CODE REVIEW

**Comment by bitbrain** [comment]
So it looks like this PR causes Godot to simply crash:
```
Run Test Suite res://test/ui/editor/property_editor/property_editor_test.gd 
	Run Test: res://test/ui/editor/property_editor/property_editor_test.gd > test_property_editor_loads :STARTED
Pandora error: value 100 is incompatible with type <RefCounted#-9223371822324513224>
Pandora error: value 42 is incompatible with type <RefCounted#-9223371822240626602>
ERROR: Resource file not found: res:// (expected type: )
   at: _load (core/io/resource_loader.cpp:279)
ERROR: Resource file not found: res:// (expected type: )
   at: _load (core/io/resource_loader.cpp:279)
ERROR: Resource file not found: res:// (expected type: )
   at: _load (core/io/resource_loader.cpp:279)
ERROR: Resource file not found: res:// (expected type: )
   at: _load (core/io/resource_loader.cpp:279)
ERROR: Resource file not found: res:// (expected type: )
   at: _load (core/io/resource_loader.cpp:279)
ERROR: Resource file not found: res:// (expected type: )
   at: _load (core/io/resource_loader.cpp:279)
ERROR: Resource file not found: res:// (expected type: )
   at: _load (core/io/resource_loader.cpp:279)
ERROR: Resource file not found: res:// (expected type: )
   at: _load (core/io/resource_loader.cpp:279)
ERROR: Resource file not found: res:// (expected type: )
   at: _load (core/io/resource_loader.cpp:279)
ERROR: Resource file not found: res:// (expected type: )
   at: _load (core/io/resource_loader.cpp:279)
ERROR: Resource file not found: res:// (expected type: )
   at: _load (core/io/resource_loader.cpp:279)
ERROR: Resource file not found: res:// (expected type: )
   at: _load (core/io/resource_loader.cpp:279)

Debugger Break, Reason: 'Invalid get index 'undefined' (on base: 'Dictionary').'
*Frame 0 - res://addons/pandora/ui/components/property_bar/property_bar.gd:24 in function 'get_scene_by_type'
Enter "help" for assistance.
debug> 
Debugger Break, Reason: 'Invalid call. Nonexistent function 'instantiate' in base 'Nil'.'
*Frame 0 - res://addons/pandora/ui/editor/property_editor/property_editor.gd:56 in function 'set_entity'
Enter "help" for assistance.
debug> Finallize .. done
```
In this particular test:
```
test_property_editor_loads 
```
I run this branch locally in 4.2 and could observe errors, so it looks like this change is breaking. things.

File: addons/pandora/model/type.gd [hunk 1]
```diff
@@ -73,8 +73,12 @@ static func lookup(name: String) -> PandoraPropertyType:
 
 	if ResourceLoader.exists(type_path):
 		var ScriptType = load(type_path)
-		if ScriptType != null and ScriptType.has_source_code():
-			return ScriptType.new()
+		if ScriptType != null:
+			if Engine.is_editor_hint() and ScriptType.has_source_code():
+				return ScriptType.new()
+		else:
+			if ScriptType.can_instantiate():
```
Comment (bitbrain) [inline addons/pandora/model/type.gd]: This will cause a nullpointer.

File: addons/pandora/model/type.gd [hunk 2]
```diff
@@ -73,8 +73,12 @@ static func lookup(name: String) -> PandoraPropertyType:
 
 	if ResourceLoader.exists(type_path):
 		var ScriptType = load(type_path)
-		if ScriptType != null and ScriptType.has_source_code():
-			return ScriptType.new()
+		if ScriptType != null:
+			if Engine.is_editor_hint() and ScriptType.has_source_code():
```
Comment (bitbrain) [inline addons/pandora/model/type.gd]: Why only when editor hint? Pandora always needs to run, regardless of editor or not.

Comment (bitbrain) [reply addons/pandora/model/type.gd]: I suggest this fix instead:
```
if ScriptType != null:
   if ScriptType.has_source_code() or ScriptType.can_instantiate():
      return ScriptType.new()
```

Comment (HolonProduction) [reply addons/pandora/model/type.gd]: If it needs to run in editor the type handlers should be marked as `@tool` and just using `can_instantiate` will work.

Comment (bitbrain) [reply addons/pandora/model/type.gd]: > If it needs to run in editor

I recommend not using `@tool` unless Godot _really_ needs to access a script at Editor time.

Comment (HolonProduction) [reply addons/pandora/model/type.gd]: That's exactly what I said or meant at least. You will have to know if you _really_ need access to those type handlers at edtior time. But from what I have seen of the addon they seem to be used to serialize values in the editor. Or am I mistaken here?

Comment (bitbrain) [reply addons/pandora/model/type.gd]: @HolonProduction good point - I'd expect in that case not using `@tool` should produce some kind of error. If that's true, we should definitely use `@tool`

Comment (HolonProduction) [reply addons/pandora/model/type.gd]: Yeah not quite sure what is intended here either. But `can_instantiate` checks whether the script is a tool when in the editor. I guess instantiating a script which returns `false` could be called undefined behavior. But this should probably raise an error. But I don't have the time to make an MRP and report it at the moment.

**Review by bitbrain — APPROVED** [review]

**Pull request merged**